libs/contrib/src/components/actions/CharacterMovement.ts

```diff
@@ -1,77 +1,43 @@
-    position.x += speed * 1.414213562373095;
-    position.y -= speed * 0.7212489168102785;
-    position.x = +(position.x.toFixed(2));
-    position.y = +(position.y.toFixed(2));
-
-    const character = entity.components.renderable as Character;
-    character.setAnimation("ur");
-  },
-  "downleft": (entity: Entity) => {
-    const position = entity.components.position as Position;
-    position.x -= speed * 1.414213562373095;
-    position.y += speed * 0.7212489168102785;
-    position.x = +(position.x.toFixed(2));
-    position.y = +(position.y.toFixed(2));
-
-    const character = entity.components.renderable as Character;
-    character.setAnimation("dl");
-  },
-  "downright": (entity: Entity) => {
-    const position = entity.components.position as Position;
-    position.x += speed * 1.414213562373095;
-    position.y += speed * 0.7212489168102785;
-    position.x = +(position.x.toFixed(2));
-    position.y = +(position.y.toFixed(2));
-
-    const character = entity.components.renderable as Character;
-    character.setAnimation("dr");
-  },
+  "upleft": (entity: Entity) => setPosAndAnimation(entity, "ul", (position) => { position.x -= speed }),
+  "upright": (entity: Entity) => setPosAndAnimation(entity, "ur", (position) => { position.y -= speed }),
+  "downleft": (entity: Entity) => setPosAndAnimation(entity, "dl", (position) => { position.y += speed }),
+  "downright": (entity: Entity) => setPosAndAnimation(entity, "dr", (position) => { position.x += speed }),
   "up": (entity: Entity) => {
-    const position = entity.components.position as Position;
-    position.y -= speed
-
-    const character = entity.components.renderable as Character;
-    character.setAnimation("u");
+    setPosAndAnimation(entity, "u", (position) => {
+      const screenXY = position.toScreenXY();
+      position.fromScreenXY(screenXY.x, screenXY.y - speed);
+    });
   },
   "down": (entity: Entity) => {
-    const position = entity.components.position as Position;
-    position.y += speed
-
-    const character = entity.components.renderable as Character;
-    character.setAnimation("d");
+    setPosAndAnimation(entity, "d", (position) => {
+      const screenXY = position.toScreenXY();
+      position.fromScreenXY(screenXY.x, screenXY.y + speed);
+    });
   },
   "left": (entity: Entity) => {
-    const position = entity.components.position as Position;
-    position.x -= speed;
-
-    const character = entity.components.renderable as Character;
-    character.setAnimation("l");
+    setPosAndAnimation(entity, "l", (position) => {
+      const screenXY = position.toScreenXY();
+      position.fromScreenXY(screenXY.x - speed, screenXY.y);
+    });
   },
   "right": (entity: Entity) => {
-    const position = entity.components.position as Position;
-    position.x += speed;
+    setPosAndAnimation(entity, "r", (position) => {
+      const screenXY = position.toScreenXY();
+      position.fromScreenXY(screenXY.x + speed, screenXY.y);
+    });
+  },
+}
```

Comment:
CharacterMovement now fully implemented in terms of incrementing world/screen coordinates